[SRVCOM-2368] Serverless multitenant istio docs

rh-max opened the pull request:
Version(s):
`serverless-docs-1.30`+

Issue:
https://issues.redhat.com/browse/SRVCOM-2368

Link to docs preview:
fixme

QE review:
- [ ] QE has approved this change.

ReToCode commented:
/cc @skonto

ReToCode commented:
/cc @pierDipi

ReToCode commented on modules/serverless-ossm-traffic-isolation-architecture.adoc:
```diff
@@ -0,0 +1,43 @@
+[id="serverless-ossm-traffic-isolation-architecture-serving"]
+== Knative Serving
+
+The Knative Serving adheres to the following architecture:
```
Maybe drop the "the". "The Knative Serving" sounds a bit strange.

ReToCode commented on modules/serverless-ossm-traffic-isolation-architecture.adoc:
```diff
@@ -0,0 +1,43 @@
+[id="serverless-ossm-traffic-isolation-architecture-eventing"]
+== Knative Eventing
+
+The Knative Serving adheres to the following architecture:
```
Eventing, maybe also without "the"?

ReToCode commented on modules/serverless-ossm-traffic-isolation-securing-the-service-mesh.adoc:
```diff
@@ -0,0 +1,257 @@
+.Example configuration for `team alpha`
+[source,terminal]
+----
+$ helm template oci://quay.io/openshift-knative/knative-istio-authz-onboarding --version 1.31.0 --set "name=team-alpha" --set "namespaces={team-alpha-1,team-alpha-2}" > team-alpha.yaml
```
There will be a productized version of the helm chart. So this probably needs to change to that URL. cc @Kaustubh-pande

ReToCode replied:
Also ideally, the version (1.31.0) is from a parameter, so that it automatically matches the version we ship. Otherwise you'll have to update this on every release.

mgencur replied:
Now that the helm chart was published, I think the right way to install it is:
```
$ helm repo add openshift-helm-charts https://charts.openshift.io/
$ helm template openshift-helm-charts/redhat-knative-istio-authz --version 1.31.0 --set "name=team-alpha" --set "namespaces={team-alpha-1,team-alpha-2}" > team-alpha.yaml
```
The first command will be done only once. The second command will be also called later for "team bravo"

The `oci://` stuff was used for developemtn purporses.

mgencur replied:
We use/test it here: https://github.com/openshift-knative/serverless-operator/blob/main/hack/generate/mesh-auth-policies.sh#L29

ReToCode reviewed (commented):
Thanks, looks very good overall. Some small things inline.

mgencur commented on integrations/serverless-ossm-setup.adoc:
```diff
@@ -6,28 +6,39 @@ include::_attributes/common-attributes.adoc[]
 [IMPORTANT]
 ====
-{ServerlessProductName} only supports the use of {SMProductName} functionality that is explicitly documented in this guide, and does not support other undocumented features.
+Using {ServerlessProductShortName} with {SMProductShortName} is only supported with {SMProductShortName} version 2.0.5 or later.
```
Yes. 2.2 and newer.

rh-max replied:
I kept this note (with the updated 2.2 version), but I stated that "2.2" applies only to 1.31 and gave the link to the article. Let's keep the note informative at least for one release :man_shrugging:

mgencur commented on modules/serverless-ossm-installing-and-configuring-openshift-service-mesh.adoc:
```diff
@@ -0,0 +1,148 @@
+        - "*"
+      tls:
+        mode: SIMPLE
+        credentialName: <wildcard_certs> <1>
```
I guess it is not clear how to generate the wildcard certificate. We remove some info about the process earlier [in this PR](https://github.com/openshift/openshift-docs/pull/68618/files#diff-32ad31e2767596655f95c2359bda26fa82b65db166ff11126f9ae567f20b61a5L23) and I can't find it anymore. Was it moved somewhere?

ReToCode replied:
Good point, we should probably keep that block (no I don't think we moved it somewhere, I just did not add it in the draft docs).

mgencur commented on modules/serverless-ossm-traffic-isolation-architecture.adoc:
```diff
@@ -0,0 +1,43 @@
+
+. Workloads in project `tenant-2` can not call workloads in project `tenant-1` directly. Requests are rejected in the Istio proxies of `tenant-1` workloads.
+
+. Workloads in project `tenant-2` can not call workloads in project `tenant-1` through `ingress-gateway`. Requests are rejected in the Istio proxies of `tenant-1` workloads.
```
Should we use `istio-ingressgateway` instead of `ingress-gateway`? When it's enclosed in apostrophes it should probably mean some real component. But nothing like `ingress-gateway` exists. Or we could use just "ingress gateway" so that we speak more generally.

mgencur reviewed (commented):
Hey @rh-max , nice stuff! 
A few comments here and there but overall looks great.

ReToCode commented on modules/serverless-init-containers-apps.adoc:
```diff
@@ -1,47 +0,0 @@
-// Module included in the following assemblies:
```
Is the removal of multiple blocks an intended change?

rh-max replied:
@ReToCode It is true that some are not related to the topic of this PR, I should probably remove them separately. If you are curious, the selection of modules I deleted is the one returned by the script for finding unused modules. (`repo/scripts/find_unused.py`)

ReToCode commented on modules/serverless-ossm-enabling-serving-metrics.adoc:
```diff
@@ -1,72 +0,0 @@
-
-:_content-type: PROCEDURE
-[id="serverless-ossm-enabling-serving-metrics_{context}"]
-= Enabling Knative Serving metrics when using Service Mesh with mTLS
```
I don't think we should drop this?

ReToCode commented on modules/serverless-ossm-secret-filtering-net-istio.adoc:
```diff
@@ -1,58 +0,0 @@
-
-.Prerequisites
-
-* You have cluster administrator permissions on {ocp-product-title}, or you have cluster or dedicated administrator permissions on {rosa-product-title} or {dedicated-product-title}.
```
Same here, I don't think we want to drop this page? @skonto ?

ReToCode commented on modules/serverless-ossm-setup-with-kourier.adoc:
```diff
@@ -1,90 +0,0 @@
-
-.Prerequisites
-
-* You have cluster administrator permissions on {ocp-product-title}, or you have cluster or dedicated administrator permissions on {rosa-product-title} or {dedicated-product-title}.
```
We will drop this but with https://issues.redhat.com/browse/SRVKS-1114. So we need to deprecate that in 1.31 first, then drop it later.

mgencur commented on modules/serverless-ossm-installing-and-configuring-openshift-service-mesh.adoc:
```diff
@@ -139,6 +139,11 @@ spec:
++
+[NOTE]
+====
+You must configure the wildcard certificate to match the domain of your {ocp-product-title} cluster. For example, if your {ocp-product-title} console address is `https://console-openshift-console.apps.openshift.example.com`, you must configure the wildcard certificate so that the domain is `*.apps.openshift.example.com`.
```
@rh-max Thanks for this. We probably still want to put back the note that we removed:
```
For more information about configuring wildcard certificates, see the following topic about _Creating a certificate to encrypt incoming external traffic_.
 ...
 include::modules/serverless-ossm-external-certs.adoc[leveloffset=+1]
```
And also put back the file serverless-ossm-external-certs.adoc which you're removing in this PR.
Having this example is quite important. Generating the certificates needs a lot of specific parameters so it's hart to figure it out without any example.

rh-max replied:
@mgencur Thanks. I have removed this particular instance of the note, because I re-added the note in the "Prerequisites" section at the top of the page (and it refers to the "Creating a certificate to encrypt incoming external traffic" section, which follows right after).

rh-max commented:
@ReToCode @mgencur  I removed the whole "Integrating Service Mesh with OpenShift Serverless" page (including the specific sections that you've mentioned):

  https://docs.openshift.com/serverless/1.30/integrations/serverless-ossm-setup.html

based on the fourth bullet point in the description in https://issues.redhat.com/browse/SRVCOM-2368 .

Do your latest comments mean that we want to keep the original section alongside the new one (at least for now)?

If so, I will put it back.

However, can you make the distinction between the two, so that I can change the names and introductory information of one of the sections? Currently they are about solving the same problem, and they even have exactly the same name:

[old](https://docs.openshift.com/serverless/1.30/integrations/serverless-ossm-setup.html) "Integrating Service Mesh with OpenShift Serverless"
[new](https://68618--docspreview.netlify.app/openshift-serverless/latest/integrations/serverless-ossm-setup) "Integrating Service Mesh with OpenShift Serverless"

Thanks.

rh-max commented:
> > @ReToCode @mgencur I removed the whole "Integrating Service Mesh with OpenShift Serverless" page (including the specific sections that you've mentioned):
> 
> I see the page title has the same name as the section in the existing page: https://docs.openshift.com/serverless/1.30/integrations/serverless-ossm-setup.html#serverless-ossm-setup_serverless-ossm-setup. I think I tried to say, to replace the linked section, not the whole page. Sorry about that confusion.
> 
> About naming. I think the page title "Integrating Service Mesh with OpenShift Serverless" is good. The section title is more about the concrete actions to set up the integration, so maybe something like "Configuring the Service Mesh integration"?

Ah, thanks for clarification Reto. I have reverted the deletion of those sections on the "Integrating Service Mesh with OpenShift Serverless" page (apart from the "Integrating Service Mesh with OpenShift Serverless" section inside it).

I have also re-added the content from the original assembly (~ introductory text at the top of the page, up until the actual steps in the instructions).

Could you @ReToCode @mgencur please check the updated text up to the section "Creating a certificate to encrypt incoming external traffic"? I want to make sure the combined text works.

mgencur commented:
> Could you @ReToCode @mgencur please check the updated text up to the section "Creating a certificate to encrypt incoming external traffic"? I want to make sure the combined text works.

@rh-max This part looks good to me.

ReToCode commented:
Looks good to me as well 👍

rh-max commented:
@ReToCode @mgencur Great, thanks for your reviews folks!

lpettyjo commented on integrations/serverless-ossm-setup.adoc:
```diff
@@ -8,14 +8,25 @@ toc::[]
+
+* All Knative internal components, as well as Knative Services, are part of the {SMProductShortName} and have sidecars injection enabled. This means that strict mTLS is enforced within the whole mesh. All requests to Knative Services require an mTLS connection, with the client having to send its certificate, except calls coming from OpenShift Routing.
+
+* {ServerlessProductName} with {SMProductShortName} integration can only target *one* service mesh. Multiple meshes can be present in the cluster, but {ServerlessProductName} will only be available on one of them.
```
Use present tense: "... is only available on one of them."

lpettyjo commented on integrations/serverless-ossm-setup.adoc:
```diff
@@ -24,10 +35,31 @@ To complete and verify these procedures in your deployment, you need either a ce
+====
+{ServerlessProductName} only supports the use of {SMProductName} functionality that is explicitly documented in this guide, and does not support other undocumented features.
+
+Using {ServerlessProductShortName} 1.31 with {SMProductShortName} is only supported with {SMProductShortName} version 2.2 or later. See the "Red Hat OpenShift Serverless Supported Configurations" page for details and information on versions other than 1.31.
```
Recast to give the reason for the link first. Also, use IBM standard language: "For information about versions other than 1.31, see the "Red Hat OpenShift Serverless Supported Configurations" page ."

lpettyjo commented on integrations/serverless-ossm-traffic-isolation.adoc:
```diff
@@ -0,0 +1,33 @@
+[id="prerequisites_serverless-ossm-traffic-isolation"]
+== Prerequisites
+
+* You have access to an {product-title} account with cluster administrator access.
```
User the the lead-in phrase "You have:" and then remove that from the bullet point text to reduce redundancy.

lpettyjo commented on modules/serverless-ossm-installing-and-configuring-openshift-service-mesh.adoc:
```diff
@@ -0,0 +1,148 @@
+<1> Enforce strict mTLS in the mesh. Only calls using a valid client certificate are allowed.
+<2> {ServerlessProductShortName} has a graceful termination for Knative Services of 30 seconds. `istio-proxy` needs to have a longer termination duration to make sure no requests are dropped.
+<3> Define a specific selector for the ingress gateway to target only the Knative gateway.
+<4> Those ports are called by Kubernetes and cluster monitoring which are not part of the mesh and cannot be called using mTLS. Hence, these ports are excluded from the mesh.
```
Rewrite as: "These ports are called by Kubernetes and cluster monitoring, which are not part of the mesh and cannot be called using mTLS. Therefore, these ports are excluded from the mesh."

lpettyjo commented on modules/serverless-ossm-traffic-isolation-architecture.adoc:
```diff
@@ -0,0 +1,43 @@
+[id="serverless-ossm-traffic-isolation-architecture_context"]
+= High-level architecture
+
+The high-level architecture of {ServerlessProductShortName} traffic isolation provided by {SMProductShortName} consists of `AuthorizationPolicies` in the `knative-serving`, `knative-eventing` and the tenants namespaces, with all the components being part of the {SMProductShortName}. The injected {SMProductShortName} sidecars take care of enforcing those rules to isolate network traffic between te ...
```
Rewrite to reduce wordiness: "The injected {SMProductShortName} sidecars enforce those rules to isolate network traffic between tenants."

lpettyjo commented on modules/serverless-ossm-traffic-isolation-architecture.adoc:
```diff
@@ -0,0 +1,43 @@
+
+The following relationships apply:
+
+. Workloads in project `tenant-2` can not call workloads in project `tenant-1` directly. Requests are rejected in the Istio proxies of `tenant-1` workloads.
+
+. Workloads in project `tenant-2` can not call workloads in project `tenant-1` through ingress gateway. Requests are rejected in the Istio proxies of `tenant-1` workloads.
+
+. Workloads in project `tenant-2` can not call workloads in project `tenant-1` through activator. Requests are rejected in the Istio proxy of the activator component. Since ingress gateway is allowed to call activator, this includes requests from project `tenant-2` through ingress gateway to activator.
```
Use "cannot" rather than "can not".

lpettyjo commented on modules/serverless-ossm-traffic-isolation-architecture.adoc:
```diff
@@ -0,0 +1,43 @@
+
+. Eventing sources, triggers and subscriptions in project `tenant-1` can send events to workloads in project `tenant-1`.
+
+. Workloads in `tenant-1` project can not send events to Eventing brokers, channels, and sinks endpoints in the `tenant-2` project.
```
Same comment as above.

lpettyjo commented on modules/serverless-ossm-traffic-isolation-architecture.adoc:
```diff
@@ -0,0 +1,43 @@
+
+. Eventing sources, triggers and subscriptions in project `tenant-2` can not send events to workloads in project `tenant-1`.
+
+. Workloads in project `tenant-1` can not call workloads in project `tenant-2` as well as workloads in project `tenant-2` can not call workloads in project `tenant-1`.
```
Same comment as above.

lpettyjo commented on modules/serverless-ossm-traffic-isolation-securing-the-service-mesh.adoc:
```diff
@@ -0,0 +1,263 @@
+* Allow traffic from `istio-system` and `knative-serving` to activator
+* Allow traffic from `knative-serving` to autoscaler
+* Allow health probes for Apache Kafka components in `knative-eventing`
+* Allow internal traffic for channel based brokers in `knative-eventing`
```
"channel-based"

lpettyjo commented on modules/serverless-ossm-traffic-isolation-securing-the-service-mesh.adoc:
```diff
@@ -0,0 +1,263 @@
++
+--
+* One `AuthorizationPolicy` limiting directly incoming traffic to the tenants OpenShift project
+* One `AuthorizationPolicy` limiting incoming traffic via the activator component of OpenShift Serverless that runs in the `knative-serving` OpenShift project
```
Use "using" rather than "via" in this instance. See IBM SG.

lpettyjo commented on modules/serverless-ossm-traffic-isolation-securing-the-service-mesh.adoc:
```diff
@@ -0,0 +1,263 @@
+* One `AuthorizationPolicy` allowing Kubernetes to call `PreStopHooks` on Knative Services
+--
++
+Instead of creating these policies by hand, install the `helm` utility and create the necessary resources for each tenant:
```
Use "manually" rather than "by hand".

lpettyjo commented on modules/serverless-ossm-verifying-installation-prerequisites.adoc:
```diff
@@ -0,0 +1,51 @@
+[id="serverless-ossm-verifying-installation-prerequisites_{context}"]
+= Verifying installation prerequisites
+
+Before installing and configuring the {SMProductShortName} integration with {ServerlessProductShortName}, you need to verify that the prerequisites have been met.
```
You can omit "you need to".

lpettyjo commented on modules/serverless-ossm-verifying-the-integration.adoc:
```diff
@@ -0,0 +1,66 @@
++
+[IMPORTANT]
+====
+You always need to add the annotation from this example to all your Knative Service to make them work with {SMProductShortName}.
```
Rewrite to reduce wordiness: "Always add the annotation from this example to all of your Knative Service..."

lpettyjo reviewed (commented):
There are multiple commits for this PR. For OCP, we squash to one commit. Not sure what your rules are. Otherwise, LGTM! Nice work!

rh-max commented:
/label merge-review-needed

adellape commented on integrations/serverless-ossm-traffic-isolation.adoc:
```diff
@@ -0,0 +1,35 @@
+[id="prerequisites_serverless-ossm-traffic-isolation"]
+== Prerequisites
+
+You have:
+
+* access to an {product-title} account with cluster administrator access
+
+* set up the {SMProductShortName} and {ServerlessProductShortName} integration
+
+* created one or more OpenShift projects for each tenant
```
I realize this was changed during peer review, but I wanted to highlight for consideration that the IBM SG does have a section about lead-in sentences that states "Do not continue the lead-in sentence after the list".

adellape commented on modules/serverless-ossm-installing-and-configuring-openshift-service-mesh.adoc:
```diff
@@ -0,0 +1,148 @@
+$ oc apply -f servicemesh-member-roll.yaml
+----
+
+. Create the necessary gateways so that {SMProductShortName} can accept traffic. The following example uses `knative-local-gateway` object using ISTIO_MUTUAL (mTLS):
```
Markup "ISTIO_MUTUAL". Also seems like a few words might be missing. Suggestion, if still technically accurate:
```suggestion
. Create the necessary gateways so that {SMProductShortName} can accept traffic. The following example uses the `knative-local-gateway` object with the `ISTIO_MUTUAL` mode (mTLS):
```

adellape commented on modules/serverless-ossm-installing-and-configuring-openshift-service-mesh.adoc:
```diff
@@ -0,0 +1,148 @@
+     targetPort: 8081
+----
+<1> Name of the secret containing the wildcard certificate.
+<2> The `knative-local-gateway` serves HTTPS traffic and expects all clients to send requests using mTLS. This means that only traffic coming from within {SMProductShortName} is possible. Workloads from outside the {SMProductShortName} must use the external domain via OpenShift Routing.
```
```suggestion
<2> The `knative-local-gateway` object serves HTTPS traffic and expects all clients to send requests using mTLS. This means that only traffic coming from within {SMProductShortName} is possible. Workloads from outside the {SMProductShortName} must use the external domain via OpenShift Routing.
```

adellape commented on modules/serverless-ossm-traffic-isolation-architecture.adoc:
```diff
@@ -0,0 +1,43 @@
+[id="serverless-ossm-traffic-isolation-architecture_context"]
+= High-level architecture
+
+The high-level architecture of {ServerlessProductShortName} traffic isolation provided by {SMProductShortName} consists of `AuthorizationPolicies` in the `knative-serving`, `knative-eventing` and the tenants namespaces, with all the components being part of the {SMProductShortName}. The injected {SMProductShortName} sidecars enforce those rules to isolate network traffic between tenants.
```
```suggestion
The high-level architecture of {ServerlessProductShortName} traffic isolation provided by {SMProductShortName} consists of `AuthorizationPolicy` objects in the `knative-serving`, `knative-eventing`, and the tenants' namespaces, with all the components being part of the {SMProductShortName}. The injected {SMProductShortName} sidecars enforce those rules to isolate network traffic between tenants.
```
Not sure where the apostrophe should go (singular tenant vs plural tenants).

adellape commented on modules/serverless-ossm-traffic-isolation-architecture.adoc:
```diff
@@ -0,0 +1,43 @@
+
+. Workloads in project `tenant-2` cannot call workloads in project `tenant-1` through ingress gateway. Requests are rejected in the Istio proxies of `tenant-1` workloads.
+
+. Workloads in project `tenant-2` cannot call workloads in project `tenant-1` through activator. Requests are rejected in the Istio proxy of the activator component. Since ingress gateway is allowed to call activator, this includes requests from project `tenant-2` through ingress gateway to activator.
```
```suggestion
. Workloads in project `tenant-2` cannot call workloads in project `tenant-1` through activator. Requests are rejected in the Istio proxy of the activator component. Because ingress gateways are allowed to call the activator, this includes requests from project `tenant-2` through ingress gateway to the activator.
```
"Because" instead of "Since" is better for translation. Also "ingress gateway is allowed" singular sounded strange, and plain "activator" without a "the" sounds strange, but disregard both if they're intentional.

adellape commented on modules/serverless-ossm-traffic-isolation-architecture.adoc:
```diff
@@ -0,0 +1,43 @@
+
+. Workloads in project `tenant-2` cannot call workloads in project `tenant-1` through activator. Requests are rejected in the Istio proxy of the activator component. Since ingress gateway is allowed to call activator, this includes requests from project `tenant-2` through ingress gateway to activator.
+
+. `AuthorizationPolicies` are applied in this project
```
```suggestion
. `AuthorizationPolicy` objects are applied in this project.
```

adellape commented on modules/serverless-ossm-traffic-isolation-architecture.adoc:
```diff
@@ -0,0 +1,43 @@
+
+. Workloads in `tenant-1` project can send events to Eventing brokers, channels, and sinks endpoints in the `tenant-1` project.
+
+. Eventing sources, triggers and subscriptions in project `tenant-1` can send events to workloads in project `tenant-1`.
```
```suggestion
. Eventing sources, triggers, and subscriptions in project `tenant-1` can send events to workloads in the `tenant-1` project.
```
Moved "project" for consistency with the other bullets.

adellape commented on modules/serverless-ossm-traffic-isolation-architecture.adoc:
```diff
@@ -0,0 +1,43 @@
+
+. Eventing sources, triggers and subscriptions in project `tenant-1` can send events to workloads in project `tenant-1`.
+
+. Workloads in `tenant-1` project cannot send events to Eventing brokers, channels, and sinks endpoints in the `tenant-2` project.
```
```suggestion
. Workloads in the `tenant-1` project cannot send events to Eventing brokers, channels, and sinks endpoints in the `tenant-2` project.
```

adellape commented on modules/serverless-ossm-traffic-isolation-architecture.adoc:
```diff
@@ -0,0 +1,43 @@
+
+. Workloads in `tenant-1` project cannot send events to Eventing brokers, channels, and sinks endpoints in the `tenant-2` project.
+
+. Eventing sources, triggers and subscriptions in project `tenant-2` cannot send events to workloads in project `tenant-1`.
```
```suggestion
. Eventing sources, triggers, and subscriptions in the `tenant-2` project cannot send events to workloads in `tenant-1` project.
```

adellape commented on modules/serverless-ossm-traffic-isolation-architecture.adoc:
```diff
@@ -0,0 +1,43 @@
+
+. Eventing sources, triggers and subscriptions in project `tenant-2` cannot send events to workloads in project `tenant-1`.
+
+. Workloads in project `tenant-1` cannot call workloads in project `tenant-2` as well as workloads in project `tenant-2` cannot call workloads in project `tenant-1`.
```
Same re: move all of the "project"s.

adellape commented on modules/serverless-ossm-traffic-isolation-architecture.adoc:
```diff
@@ -0,0 +1,43 @@
+
+. Workloads in project `tenant-1` cannot call workloads in project `tenant-2` as well as workloads in project `tenant-2` cannot call workloads in project `tenant-1`.
+
+. `AuthorizationPolicies` are applied in this project.
```
```suggestion
. `AuthorizationPolicy` objects are applied in this project.
```

adellape commented on modules/serverless-ossm-traffic-isolation-securing-the-service-mesh.adoc:
```diff
@@ -0,0 +1,263 @@
+[id="serverless-ossm-traffic-isolation-securing-the-service-mesh_{context}"]
+= Securing the {SMProductShortName}
+
+Authorization Policies and mTLS allow you to secure {SMProductShortName}.
```
```suggestion
Authorization policies and mTLS allow you to secure {SMProductShortName}.
```
Unless you want it to be "`AuthorizationPolicy` objects".

adellape commented on modules/serverless-ossm-traffic-isolation-securing-the-service-mesh.adoc:
```diff
@@ -0,0 +1,263 @@
+   - team-bravo-2       # example OpenShift project that belongs th the team-bravo tenant
+----
++
+All OpenShift projects that are part of the mesh must enforce mTLS in strict mode. This forces Istio to only accept connections with a client-certificate present and allows the Service Mesh sidecar to validate the origin using `AuthorizationPolicy`.
```
```suggestion
All OpenShift projects that are part of the mesh must enforce mTLS in strict mode. This forces Istio to only accept connections with a client-certificate present and allows the Service Mesh sidecar to validate the origin using an `AuthorizationPolicy` object.
```

adellape commented on modules/serverless-ossm-traffic-isolation-securing-the-service-mesh.adoc:
```diff
@@ -0,0 +1,263 @@
++
+All OpenShift projects that are part of the mesh must enforce mTLS in strict mode. This forces Istio to only accept connections with a client-certificate present and allows the Service Mesh sidecar to validate the origin using `AuthorizationPolicy`.
+
+. Create the configuration with `AuthorizationPolicies` in the `knative-serving` and the `knative-eventing` namespace:
```
```suggestion
. Create the configuration with `AuthorizationPolicy` objects in the `knative-serving` and `knative-eventing` namespaces:
```

adellape commented on modules/serverless-ossm-traffic-isolation-securing-the-service-mesh.adoc:
```diff
@@ -0,0 +1,263 @@
+These policies restrict the access rules for the network communication between {ServerlessProductShortName} system components. Specifically, they enforce the following rules:
++
+--
+* Deny all traffic that is not explicitly allowed in `knative-serving` and `knative-eventing`
```
```suggestion
* Deny all traffic that is not explicitly allowed in the `knative-serving` and `knative-eventing` namespaces
```

adellape commented on modules/serverless-ossm-traffic-isolation-securing-the-service-mesh.adoc:
```diff
@@ -0,0 +1,263 @@
++
+--
+* Deny all traffic that is not explicitly allowed in `knative-serving` and `knative-eventing`
+* Allow traffic from `istio-system` and `knative-serving` to activator
+* Allow traffic from `knative-serving` to autoscaler
+* Allow health probes for Apache Kafka components in `knative-eventing`
+* Allow internal traffic for channel-based brokers in `knative-eventing`
```
Similarly add "the X namespace" styling where appropriate in these.

adellape commented on modules/serverless-ossm-traffic-isolation-securing-the-service-mesh.adoc:
```diff
@@ -0,0 +1,263 @@
+$ oc apply -f knative-default-authz-policies.yaml
+----
+
+. Define which OpenShift projects can communicate with each other. For that, every OpenShift project of a tenant needs:
```
```suggestion
. Define which OpenShift projects can communicate with each other. For this communication, every OpenShift project of a tenant requires the following:
```

adellape commented on modules/serverless-ossm-traffic-isolation-verifying-the-configuration.adoc:
```diff
@@ -0,0 +1,216 @@
+[id="serverless-ossm-traffic-isolation-verifying-the-configuration_{context}"]
+= Verifying the configuration
+
+You can use `curl` to verify the configuration for network traffic isolation.
```
```suggestion
You can use the `curl` command to verify the configuration for network traffic isolation.
```

adellape commented on modules/serverless-ossm-traffic-isolation-verifying-the-configuration.adoc:
```diff
@@ -0,0 +1,216 @@
+
+[NOTE]
+====
+The following examples assume having two tenants, each having one namespace, and all part of the `ServiceMeshMemberRoll` object, configured with the resources in `team-alpha.yaml` and `team-bravo.yaml`.
```
```suggestion
The following examples assume having two tenants, each having one namespace, and all part of the `ServiceMeshMemberRoll` object, configured with the resources in the `team-alpha.yaml` and `team-bravo.yaml` files.
```

adellape commented on modules/serverless-ossm-traffic-isolation-verifying-the-configuration.adoc:
```diff
@@ -0,0 +1,216 @@
+EOF
+----
+
+. Verify the configuration using `curl`.
```
```suggestion
. Verify the configuration by using the `curl` command.
```

adellape commented on modules/serverless-ossm-traffic-isolation-verifying-the-configuration.adoc:
```diff
@@ -0,0 +1,216 @@
+.Example command
+[source,terminal]
+----
+oc exec deployment/curl -n team-alpha-1 -it -- curl -v http://test-webapp.team-alpha-1:80
```
```suggestion
$ oc exec deployment/curl -n team-alpha-1 -it -- curl -v http://test-webapp.team-alpha-1:80
```

adellape commented on modules/serverless-ossm-verifying-installation-prerequisites.adoc:
```diff
@@ -0,0 +1,51 @@
++
+[NOTE]
+====
+The mesh that {ServerlessProductShortName} is part of has to be distinct and preferably reserved only for {ServerlessProductShortName} workloads. That is because additional configuration, such as `Gateways`, might interfere with the {ServerlessProductShortName} gateways `knative-local-gateway` and `knative-ingress-gateway`. {SMProductName} only allows one Gateway to claim a wildcard host binding  ...
```
```suggestion
The mesh that {ServerlessProductShortName} is part of must be distinct and preferably reserved only for {ServerlessProductShortName} workloads. That is because additional configuration, such as `Gateways`, might interfere with the {ServerlessProductShortName} gateways `knative-local-gateway` and `knative-ingress-gateway`. {SMProductName} only allows one Gateway to claim a wildcard host binding (`hosts: ["*"]`) on the same port (`port: 443`). If another Gateway is already binding this configuration, a separate mesh has to be created for {ServerlessProductShortName} workloads. 
```

adellape commented on modules/serverless-ossm-verifying-installation-prerequisites.adoc:
```diff
@@ -0,0 +1,51 @@
+istio-system   istio-ingressgateway  ClusterIP  172.30.46.146 none>   15021/TCP,80/TCP,443/TCP     9m50s
+----
++
+This command should not return a Service of type `NodePort` or `LoadBalancer`.
```
```suggestion
This command should not return a service of type `NodePort` or `LoadBalancer`.
```
or:
```suggestion
This command should not return a `Service` object of type `NodePort` or `LoadBalancer`.
```

adellape commented on modules/serverless-ossm-verifying-installation-prerequisites.adoc:
```diff
@@ -0,0 +1,51 @@
++
+[NOTE]
+====
+Cluster external Knative Services are expected to be called via OpenShift Ingress using OpenShift Routes. It is not supported to access {SMProductShortName} directly, such as by exposing the `istio-ingressgateway` using a `Service` with type `NodePort` or `LoadBalancer`.
```
```suggestion
Cluster external Knative Services are expected to be called via OpenShift Ingress using OpenShift Routes. It is not supported to access {SMProductShortName} directly, such as by exposing the `istio-ingressgateway` using a `Service` object with type `NodePort` or `LoadBalancer`.
```

adellape commented on modules/configuring-sinkbinding-with-ossm.adoc:
```diff
@@ -0,0 +1,159 @@
+$ oc apply -f heartbeat-sinkbinding.yaml
+----
+
+. Create a `CronJob`:
```
```suggestion
. Create a `CronJob` object:
```

adellape commented on modules/configuring-sinkbinding-with-ossm.adoc:
```diff
@@ -0,0 +1,159 @@
+$ oc apply -f event-display-service.yaml
+----
+
+. Create a `SinkBinding`:
```
```suggestion
. Create a `SinkBinding` object:
```

adellape commented on modules/configuring-sinkbinding-with-ossm.adoc:
```diff
@@ -0,0 +1,159 @@
+
+.Procedure
+
+. Create a `Service` in a namespace that is member of the `ServiceMeshMemberRoll`:
```
```suggestion
. Create a `Service` object in a namespace that is member of the `ServiceMeshMemberRoll`:
```

adellape commented on modules/configuring-sinkbinding-with-ossm.adoc:
```diff
@@ -0,0 +1,159 @@
+<1> A namespace that is a member of the `ServiceMeshMemberRoll`.
+<2> This annotation injects {SMProductShortName} sidecars into the Knative service pods.
+
+. Apply the `Service` resource:
```
```suggestion
. Apply the `Service` object:
```
Just for consistency with other parts of the PR, though switching everything from "object" to "resource" is also fine.

adellape commented on modules/configuring-sinkbinding-with-ossm.adoc:
```diff
@@ -0,0 +1,159 @@
+<1> A namespace that is part of the `ServiceMeshMemberRoll`.
+<2> Bind any Job with the label `app: heartbeat-cron` to the event sink.
+
+. Apply the `SinkBinding` resource:
```
```suggestion
. Apply the `SinkBinding` object:
```

adellape commented on modules/configuring-sinkbinding-with-ossm.adoc:
```diff
@@ -0,0 +1,159 @@
+<1> A namespace that is part of the `ServiceMeshMemberRoll`.
+<2> Inject Service Mesh sidecars into the `CronJob` pods.
+
+. Apply the `CronJob` resource:
```
```suggestion
. Apply the `CronJob` object:
```

adellape commented on modules/serverless-ossm-installing-and-configuring-openshift-serverless.adoc:
```diff
@@ -0,0 +1,201 @@
+$ oc apply -f knative-serving-config.yaml
+----
+
+. Install Knative Eventing with the following `KnativeEventing` custom resource, which enables the Istio integration:
```
```suggestion
. Install Knative Eventing with the following `KnativeEventing` object, which enables the Istio integration:
```
Again, just for consistency. "Custom resource" isn't really necessary to call out here, I don't think.

adellape commented on modules/serverless-ossm-installing-and-configuring-openshift-serverless.adoc:
```diff
@@ -0,0 +1,201 @@
+<1> The Apache Kafka cluster URL, for example `my-cluster-kafka-bootstrap.kafka:9092`.
+<2> Enable sidecar injection for Knative Kafka pods.
+
+. Apply the `KnativeEventing` resource:
```
```suggestion
. Apply the `KnativeEventing` object:
```
I'll stop highlighting any "resource" vs "object" discrepancies, and leave it to you to decide which to stick with throughout the PR.

adellape commented on modules/serverless-ossm-installing-and-configuring-openshift-service-mesh.adoc:
```diff
@@ -0,0 +1,148 @@
++
+[IMPORTANT]
+====
+If you have an existing `ServiceMeshControlPlane`, make sure that you have the same configuration applied.
```
```suggestion
If you have an existing `ServiceMeshControlPlane` object, make sure that you have the same configuration applied.
```

adellape reviewed (commented):
Dropping the merge-review labels for some issues that I feel should be addressed first after checking with our DPM (namely the diagrams issue).

Regarding the GA timing, note for the future that we also have a [new internal process](https://docs.google.com/document/d/1fLMpK4bqthtFlCwA36yeo2cI-pExFpX76VMnSi_In6Q/edit#heading=h.x8ydklaezdcq) for requesting a pre-merge merge-review when we know about an upcoming release date that a PR should be timed with, in case that's helpful.

rh-max commented on integrations/serverless-ossm-traffic-isolation.adoc:
```diff
@@ -0,0 +1,35 @@
+:FeatureName: Using {SMProductShortName} to isolate network traffic with {ServerlessProductName}
+include::snippets/technology-preview.adoc[leveloffset=+2]
+
+{SMProductShortName} can be used to isolate network traffic between tenants on a shared OpenShift cluster using Service Mesh `AuthorizationPolicy` resources. {ServerlessProductShortName} can also leverage this, using several {SMProductShortName} resources. A tenant is a group of one or multiple OpenShift projects that can access each other over the network on a shared OpenShift cluster.
```
Went through all instances of plaintext "OpenShift" and changed/removed where appropriate.

rh-max commented on images/multi-tenancy-service-mesh-eventing.png:
Requested the diagrams and removed them temporarily. The SME approved the go-ahead without the sections with the diagrams, so we want to go ahead and merge this. The sections with new diagrams will be re-added in a future PR (once diagrams are available).

adellape commented on modules/serverless-ossm-installing-and-configuring-openshift-serverless.adoc:
```diff
@@ -0,0 +1,201 @@
+----
+<1> Enable Istio integration.
+<2> Enable sidecar injection for Knative Serving data plane pods.
+<3> If your istio is not running in the `istio-system` namespace, you need to set these two flags with the correct namespace.
```
For a follow-up:
```suggestion
<3> If your Istio is not running in the `istio-system` namespace, you must set these two flags with the correct namespace.
```
I think re: capitalization?

adellape commented on modules/serverless-ossm-installing-and-configuring-openshift-serverless.adoc:
```diff
@@ -0,0 +1,201 @@
+$ oc apply -f knative-kafka-config.yaml
+----
+
+. Install `ServiceEntry` to inform {SMProductShortName} of the communication between `KnativeKafka` components and an Apache Kafka cluster:
```
For a follow-up:
```suggestion
. Create a `ServiceEntry` object to inform {SMProductShortName} of the communication between `KnativeKafka` components and an Apache Kafka cluster:
```
Maybe?

adellape commented on modules/serverless-ossm-installing-and-configuring-openshift-service-mesh.adoc:
```diff
@@ -0,0 +1,148 @@
+<3> Define a specific selector for the ingress gateway to target only the Knative gateway.
+<4> These ports are called by Kubernetes and cluster monitoring, which are not part of the mesh and cannot be called using mTLS. Therefore, these ports are excluded from the mesh.
+
+. Add the namespaces that you would like to integrate with {SMProductShortName} to the `ServiceMeshMemberRoll` object as members:
```
For a follow-up: I noticed earlier there were a lot of "part of the `ServiceMeshMemberRoll`" or "member of the `ServiceMeshMemberRoll`" without an "object" on the end, but I didn't call out as I wasn't sure of the implication. But it looks like here that you've got "object" on the end, so suggest going back to the others and adding for consistency as well.

adellape commented on modules/serverless-ossm-traffic-isolation-securing-the-service-mesh.adoc:
```diff
@@ -0,0 +1,263 @@
+$ oc apply -f knative-default-authz-policies.yaml
+----
+
+. Define which OpenShift projects can communicate with each other. For this communication, every OpenShift project of a tenant requires the following:
```
For a follow-up: A few plain "OpenShift"s here.

adellape commented on modules/serverless-ossm-traffic-isolation-securing-the-service-mesh.adoc:
```diff
@@ -0,0 +1,263 @@
++
+Instead of creating these policies manually, install the `helm` utility and create the necessary resources for each tenant:
++
+.Installing the `helm` utility
+[source,terminal]
+----
+$ helm repo add openshift-helm-charts https://charts.openshift.io/
+----
++
+.Creating example configuration for `team alpha`
+[source,terminal,subs="attributes+"]
+----
+$ helm template openshift-helm-charts/redhat-knative-istio-authz --version {ServerlessProductVersion} --set "name=team-alpha" --set "namespaces={team-alpha-1,team-alpha-2}" > team-alpha.yaml
+----
++
+.Creating example configuration for `team bravo`
+[source,terminal,subs="attributes+"]
+----
+$ helm template openshift-helm-charts/redhat-knative-istio-authz --version 1.31.0 --set "name=team-bravo" --set "namespaces={team-bravo-1,team-bravo-2}" > team-bravo.yaml
+----
```
For a follow-up: Using block titles to describe (what seems like) sequential commands/steps is a bit non-standard. I'd suggest breaking these each into actual substeps.

adellape commented on modules/serverless-ossm-traffic-isolation-verifying-the-configuration.adoc:
```diff
@@ -0,0 +1,216 @@
+
+. Verify the configuration by using the `curl` command.
++
+Test `team-alpha-1 -> team-alpha-1` through cluster local domain, which is allowed:
```
For a follow-up: One left-over `->` style here. Could be:
```suggestion
Test the `team-alpha-1` to `team-alpha-1` connection through the cluster's local domain, which is allowed:
```

adellape commented on modules/serverless-ossm-traffic-isolation-verifying-the-configuration.adoc:
```diff
@@ -0,0 +1,216 @@
+RBAC: access denied
+----
++
+Test the `team-alpha-1` to `team-bravo-1` connection through an external domain, which is allowed:
```
For a follow-up: Also in hindsight / upon re-review, each of these "Test the X connection..."s should probably be substeps (maybe just unordered bullets if they're not actually required to be sequential/ordered).

adellape commented on modules/serverless-ossm-verifying-installation-prerequisites.adoc:
```diff
@@ -0,0 +1,51 @@
++
+[NOTE]
+====
+The mesh that {ServerlessProductShortName} is part of must be distinct and preferably reserved only for {ServerlessProductShortName} workloads. That is because additional configuration, such as `Gateways`, might interfere with the {ServerlessProductShortName} gateways `knative-local-gateway` and `knative-ingress-gateway`. {SMProductName} only allows one Gateway to claim a wildcard host binding (` ...
```
For a follow-up:
```suggestion
The mesh that {ServerlessProductShortName} is part of must be distinct and preferably reserved only for {ServerlessProductShortName} workloads. That is because additional configuration, such as `Gateways`, might interfere with the {ServerlessProductShortName} gateways `knative-local-gateway` and `knative-ingress-gateway`. {SMProductName} only allows one Gateway to claim a wildcard host binding (`hosts: ["*"]`) on the same port (`port: 443`). If another Gateway is already binding this configuration, a separate mesh must be created for {ServerlessProductShortName} workloads. 
```

adellape commented on modules/serverless-ossm-verifying-installation-prerequisites.adoc:
```diff
@@ -0,0 +1,51 @@
+The mesh that {ServerlessProductShortName} is part of must be distinct and preferably reserved only for {ServerlessProductShortName} workloads. That is because additional configuration, such as `Gateways`, might interfere with the {ServerlessProductShortName} gateways `knative-local-gateway` and `knative-ingress-gateway`. {SMProductName} only allows one Gateway to claim a wildcard host binding (` ...
+====
+
+. Check whether {SMProductName} `istio-ingressgateway` is exposed as type `NodePort` or `LoadBalancer`:
```
For a follow-up:
```suggestion
. Check whether the {SMProductName} `istio-ingressgateway` is exposed as type `NodePort` or `LoadBalancer`:
```

adellape commented on modules/serverless-ossm-verifying-the-integration.adoc:
```diff
@@ -0,0 +1,66 @@
+$ curl --cacert root.crt <service_url>
+----
++
+For example, run:
```
For a follow-up:
```suggestion
For example, run the following command:
```

adellape commented on _topic_maps/_topic_map.yml:
```diff
@@ -312,6 +312,10 @@ Topics:
     File: serverless-ha
 - Name: Configuring kube-rbac-proxy resources for Eventing
   File: kube-rbac-proxy-eventing
+- Name: Using ContainerSource with OpenShift Service Mesh
```
For a follow-up: Consider one of the following:
```suggestion
- Name: Using a container source with OpenShift Service Mesh
```
^ Probably preferable, to match the "Using a sink binding..." style of the following topic.

Or (note no markup because heading):
```suggestion
- Name: Using a ContainerSource object with OpenShift Service Mesh
```
If changed, apply the styling to the other 2 similar instances (the `eventing/containersource-with-ossm.adoc` assembly and the `modules/configuring-containersource-with-ossm.adoc` procedure) to match.

adellape reviewed (approved):
Thank you for the updates, it looks great! I've added some nits I noticed during re-review that could be considered for a follow-up PR, but nothing that should block merge at this time.

adellape commented:
/cherrypick serverless-docs-1.30

adellape commented:
/cherrypick serverless-docs-1.31
